core/variant/dictionary.cpp

```diff
@@ -68,27 +68,53 @@ void Dictionary::get_key_list(List<Variant> *p_keys) const {
 }
 
 Variant Dictionary::get_key_at_index(int p_index) const {
-	int index = 0;
-	for (const KeyValue<Variant, Variant> &E : _p->variant_map) {
-		if (index == p_index) {
-			return E.key;
-		}
-		index++;
+	const int n = size();
+	if (p_index < -n || p_index >= n) {
+		return Variant(); // Return null when out of bounds.
 	}
 
-	return Variant();
+	if (p_index < 0) {
+		p_index += n; // Convert negative index to positive.
+	}
+
+	if (p_index <= n / 2) {
+		HashMap<Variant, Variant, VariantHasher, StringLikeVariantComparator>::ConstIterator E(_p->variant_map.begin());
+		for (int i = 0; i < p_index; ++i) {
+			++E;
+		}
+		return E->key;
+	} else {
+		HashMap<Variant, Variant, VariantHasher, StringLikeVariantComparator>::ConstIterator E(_p->variant_map.last());
+		for (int i = n - 1; i > p_index; --i) {
+			--E;
+		}
+		return E->key;
+	}
```

Comment:
Since you're doing the same thing twice now, I think it warrants having an internal helper function like `KeyValue<Variant, Variant> *get_key_value_at_index()` and use that in both functions.

Reply:
Thanks! Done with latest commit, and this helped speed up the C# function too which was calling both `get_key_at_index` and `get_value_at_index` independently.